R/cov_struct.R

```diff
@@ -0,0 +1,289 @@
+#'
+#' @family cov_struct
+#' @export
+cov_struct <- function(type = cov_types(), visits, subject, group = NULL,
+  heterogeneous = FALSE) {
+
+  # if heterogeneous isn't provided, derive from provided type
+  if (missing(heterogeneous)) {
+    heterogeneous <- switch(type,
+      toeph = , ar1h = , adh = , csh = TRUE,
+      heterogeneous
+    )
+  }
+
+  # coerce all type options into abbreviated form
+  type <- match.arg(type)
+  type <- COV_TYPES[, 1][which(COV_TYPES == type, arr.ind = TRUE)[, 1]]
+
+  x <- structure(
+    list(
+      type = type,
+      heterogeneous = heterogeneous,
+      visits = visits,
+      subject = subject,
+      group = group
+    ),
+    class = c("cov_struct", "mmrm_cov_struct", "list")
+  )
+
+  validate_cov_struct(x)
+}
+
+
+
+#' Validate Covariance Structure Data
+#'
+#' Run checks against relational integrity of covariance definition
+#'
+#' @param x (`cov_struct`)\cr
+#'   A covariance structure object
+#'
+#' @return `x` if successful, or an error is thrown otherwise
+#'
+#' @keywords internal
+validate_cov_struct <- function(x) {
+  check("stop", length(x$subject) == 1,
+    "Exactly one variable must be provided for the covariance structure's `subject`")
+
+  check("stop", length(x$group) <= 1,
+    "At most one variable is allowed for the covariance structure's `group`")
+
+  check("stop", length(x$visits) > 0,
+    "A variable must be provided for the covariance structure's `time`")
+
+  check("stop", x$type == "sp_exp" || length(x$visits) < 2,
+    "Multiple `time` variables are only permitted for 'spatial' type covariance")
+
+  check("warn", !any(dup <- duplicated(x$visits)),
+    "Duplicated `time` variable(s): ", fmt_syms(unique(x$visits[dup])))
+
+  emit_check_results()
+  x
+}
+
+
+
+#' Format Covariance Structure Object
+#'
+#' @param x (`cov_struct`)\cr
+#'   A covariance structure object
+#' @param ...
+#'   Additional arguments unused
+#'
+#' @return A formatted string for `x`
+#'
+#' @export
+format.cov_struct <- function(x, ...) {
+  long_name <- COV_TYPES[, 3][which(COV_TYPES == x$type, arr.ind = TRUE)[, 1]]
+  sprintf("<covariance structure>\n%s%s:\n\n  %s | %s%s\n",
+    long_name,
+    if (x$heterogeneous) " (heterogeneous)" else "",
+    fmt_syms(x$visits),
+    if (!is.null(x$group)) paste0(fmt_syms(x$group), " / ") else "",
+    fmt_syms(x$subject)
+  )
+}
+
+
+
+#' Print a Covariance Structure Object
+#'
+#' @param x (`cov_struct`)\cr
+#'   A covariance structure object
+#' @param ...
+#'   Additional arguments unused
+#'
+#' @return `x` invisibly
+#'
+#' @export
+print.cov_struct <- function(x, ...) {
+  cat(format(x, ...), "\n")
+  invisible(x)
+}
+
+
+
+#' Coerce into a Covariance Structure Definition
+#'
+#' @details
+#' A covariance structure can be parsed from a model definition formula or call.
+#' Generally, covariance structures defined using non-standard evaluation take
+#' the following form:
+#'
+#' ```
+#' type( (visit, )* visit | (group /)? subject )
```

Comment:
I would separate the spatial and non-spatial covariance structures here for the explanation, because for spatial ones we talk about coordinates instead of visits, see e.g. https://github.com/openpharma/mmrm/blob/main/vignettes/algorithm.Rmd#L168